src/support/generation.c

```diff
@@ -0,0 +1,360 @@
+__wt_gen(WT_SESSION_IMPL *session, int which)
+{
+	return (S2C(session)->generations[which]);
+}
+
+/*
+ * __wt_gen_next --
+ *	Switch the resource to its next generation.
+ */
+uint64_t
+__wt_gen_next(WT_SESSION_IMPL *session, int which)
+{
+	return (__wt_atomic_addv64(&S2C(session)->generations[which], 1));
+}
+
+#if 0
+/*
+ * __wt_gen_next_drain --
+ *	Switch the resource to its next generation, then wait for it to drain.
+ */
+uint64_t
+	TABBED IN to avoid dist/ functions:
+	__wt_gen_next_drain(WT_SESSION_IMPL *session, int which)
+{
+	uint64_t v;
+
+	v = __wt_atomic_addv64(&S2C(session)->generations[which], 1);
+
+	__wt_gen_drain(session, which, v);
+
+	return (v);
+}
+
+/*
+ * __wt_gen_drain --
+ *	Wait for the resource to drain.
+ */
+void
+	TABBED IN to avoid dist/ functions:
+	__wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_SESSION_IMPL *s;
+	uint64_t v;
+	uint32_t i, session_cnt;
+	int pause_cnt;
+
+	conn = S2C(session);
+
+	/*
+	 * No lock is required because the session array is fixed size, but it
+	 * may contain inactive entries. We must review any active session, so
+	 * insert a read barrier after reading the active session count. That
+	 * way, no matter what sessions come or go, we'll check the slots for
+	 * all of the sessions that could have been active when we started our
+	 * check.
+	 */
+	WT_ORDERED_READ(session_cnt, conn->session_cnt);
+	for (pause_cnt = 0,
+	    s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
+		if (!s->active)
+			continue;
+
+		for (;;) {
+			/* Ensure we only read the value once. */
+			WT_ORDERED_READ(v, s->generations[which]);
+
+			if (v == 0 || generation <= v)
+				break;
+
+			/*
+			 * The pause count is cumulative, quit spinning if it's
+			 * not doing us any good, that can happen in generations
+			 * that don't move quickly.
+			 */
+			if (++pause_cnt < WT_THOUSAND)
+				WT_PAUSE();
+			else
+				__wt_sleep(0, 10);
+		}
+	}
+}
+#endif
+
+/*
+ * __wt_gen_oldest --
+ *	Return the oldest generation in use for the resource.
+ */
+uint64_t
+__wt_gen_oldest(WT_SESSION_IMPL *session, int which)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_SESSION_IMPL *s;
+	uint64_t oldest, v;
+	uint32_t i, session_cnt;
+
+	conn = S2C(session);
+
+	/*
+	 * No lock is required because the session array is fixed size, but it
+	 * may contain inactive entries. We must review any active session, so
+	 * insert a read barrier after reading the active session count. That
+	 * way, no matter what sessions come or go, we'll check the slots for
+	 * all of the sessions that could have been active when we started our
+	 * check.
+	 */
+	WT_ORDERED_READ(session_cnt, conn->session_cnt);
+	for (oldest = conn->generations[which] + 1,
+	    s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
+		if (!s->active)
+			continue;
+
+		/* Ensure we only read the value once. */
+		WT_ORDERED_READ(v, s->generations[which]);
+		if (v != 0 && v < oldest)
+			oldest = v;
+	}
```

Comment:
One thing I'd like to talk through with this change is whether we can avoid this loop (and in fact any loop over sessions).

Ignoring performance issues for a second, imagine each generation number is hashed to a WT_RWLOCK.  Entering generation 100 would acquire a read lock on that generation's lock.

When some thread wants to be sure that everyone has left generation 100, it first increments the generation to 101, then waits for a write lock on the lock for generation 100.  This will be a huge win when there are 100K sessions open.

I think we can make the performance reasonable by having N locks per generation (where 8 <= N <= 32).  Each session entering a generation uses `session->id` to hash onto one of these locks, making conflicts between sessions very unlikely.  There is more work for the thread checking everyone has left a generation (it has to get N writelocks serially), but still much less than scanning all the sessions.